pyanaconda/dbus/server.py

```diff
@@ -0,0 +1,459 @@
+
+    def _get_specification(self):
+        """Get the DBus specification.
+
+        :return: a DBus specification
+        """
+        return DBusSpecification.from_xml(self._get_xml_specification())
+
+    @abstractmethod
+    def _get_xml_specification(self):
+        """Get the XML specification.
+
+        :return: a XML specification
+        """
+        return ""
+
+    @abstractmethod
+    def connect_object(self):
+        """Connect the object to DBus.
+
+        Handle emitted signals of the object with the _emit_signal method
+        and handle incoming DBus calls with the _handle_call method.
+        """
+        pass
+
+    @abstractmethod
+    def disconnect_object(self):
+        """Disconnect the object from DBus.
+
+        Unregister the object and disconnect all signals.
+        """
+        pass
+
+    @abstractmethod
+    def _connect_signal(self, interface_name, signal_name):
+        """Connect a DBus signal.
+
+        :param interface_name: a DBus interface name
+        :param signal_name: a DBus signal name
+        """
+        pass
+
+    @abstractmethod
+    def _emit_signal(self, interface_name, signal_name, *parameters):
+        """Handle a DBus signal.
+
+        :param interface_name: a DBus interface name
+        :param signal_name: a DBus name of the signal
+        :param parameters: a signal parameters
+        """
+        pass
+
+    def _handle_call(self, interface_name, method_name, *parameters):
+        """Handle a DBus call.
+
+        :param interface_name: a name of the interface
+        :param method_name: a name of the called method
+        :param parameters: parameters of the call
+        :return: a result of the DBus call
+        """
+        handler = self._find_handler(interface_name, method_name)
+        return handler(*parameters)
+
+    def _find_member_spec(self, interface_name, member_name):
+        """Find a specification of the DBus member.
+
+        :param interface_name: a name of the interface
+        :param member_name: a name of the member
+        :return: a specification of the member
+        """
+        return self.specification.get_member(
+            interface_name, member_name
+        )
+
+    def _find_handler(self, interface_name, member_name):
+        """Find a handler of a DBus member.
+
+        :param interface_name: a name of the interface
+        :param member_name: a name of the method
+        :return: a handler
+        """
+        handler = self._find_object_handler(interface_name, member_name) \
+            or self._find_default_handler(interface_name, member_name)
+
+        if not handler:
+            raise AttributeError("The member {}.{} has no handler.".format(
+                interface_name, member_name
+            ))
+
+        return handler
+
+    def _find_object_handler(self, interface_name, member_name):
+        """Get an object handler of a DBus call.
+
+        :param interface_name: a name of the interface.
```

Comment:
The parameter not being used to find the handler made me doubt if I understand how picking the method of the right interface works. Is it ensured by class inheritance (and attirbute access) mapping to dbus interface inheritance in our case?

Reply:
DBus interfaces with members of the same name are not supported. Our `@dbus_interface` and `@dbus_class` decorators should fail to generate the XML specification in this case.